### PR TITLE
Use PIL as default image handler

### PR DESCRIPTION
**Pull request by flying-sheep**
The default used to be None

### CODE REVIEW

**Comment by takluyver** [comment]
Might need a try/except around the import in order to fail gracefully when PIL is not installed.

**Comment by takluyver** [comment]
Also looks like it needs  a tweak to the tests.

**Comment by Carreau** [comment]
I believe the default was None because on non-ssd-hi-end laptop PIL is relatively slow to load. 

As the same time should we replace/complement PIL by Pillow ?

**Comment by minrk** [comment]
The import for pillow is still the same, isn't it?

**Comment by takluyver** [comment]
Yeah, I think Pillow is imported as PIL.

**Comment by Carreau** [comment]
> The import for pillow is still the same, isn't it?

It seem you are right.

**Comment by flying-sheep** [comment]
that’s because you confused packages with packages, duh :wink:

the *PIP distributable* installable by the name “Pillow” contains a *python namespace* importable with the name “PIP”.

### rant incoming!

python terminology here is pretty retarded, which makes it hard to understand

* [module](https://docs.python.org/3/tutorial/modules.html): a .py file you can `import`
* [package](https://docs.python.org/3/tutorial/modules.html#packages) (1. definition): multiple
* [package](https://pypi.python.org/pypi) (2. definition): something you can install via `setup.py` or `pip`. has a name and can contain 0-∞ top-level packages (1. definition) of arbitrary names

this is so dumb. “modules” and “packages” (1. definition) are the same thing: importable namespaces, while “packages” (2. definition) are something completely different: a distributable container with install instructions for binaries and optionally *multiple* namespaces.

i propose that we start calling “packages” (1. definition)

**Comment by takluyver** [comment]
That's not entirely accurate:

- A module is an importable name. It can be a .py file or other things. An extension module is a shared library, and you can also have builtin modules and modules loaded from zip files.
- A package is a specific type of module which can also contain other modules. When you import things normally from the filesystem, a package corresponds to a directory.
- A *distribution* is a file that gets downloaded and installed. This includes source distributions which contain setup.py, and binary distributions such as wheels. We often call these 'packages', which confuses matters, but the technical term in Python is 'distributions'.

I agree that it's unfortunate that distributions do not have a 1:1 correspondence with modules - I think that way back in Python's history, distributions were defined before packages, so the only way for a distribution to include >1 file was to have multiple top level modules. This is one thing [flit](http://flit.readthedocs.io/en/latest/) changes: each distribution contains exactly one top-level module.

**Comment by flying-sheep** [comment]
however bad they are, my definitions are correct. i’m just repeating what the official docs and PyPI says. if you (like me) have a problem with the definition, we must work to change them.

here are the relevant parts:

1. > A module is a file containing Python definitions and statements. The file name is the module name with the suffix `.py` appended.
2. > Packages are a way of structuring Python’s module namespace by using “dotted module names”. […] When importing the package, Python searches through the directories on sys.path looking for the package subdirectory.
3. > The Python Package Index is a repository of software for the Python programming language. There are currently 81764 packages here. […] To use a package from this index either "pip install package" (get pip) or download, unpack and "python setup.py install" it.

**Comment by takluyver** [comment]
The technical term for installable objects is already 'distributions'. Calling them packages is widespread, but we do already have a way to distinguish them without changing any nomenclature.

**Comment by flying-sheep** [comment]
then this needs to be mentioned on the PyPI homepage and the docs. [this](https://docs.python.org/3/distributing/index.html) exclusively talks about “packages”

**Comment by Carreau** [comment]
Yes, they are many things that would need to be changed/ standardized across python packages. 
regardless, +1 to merging that as is, refine if necessary. 

I'll make 1 inline comment  which is not a request to change but a question.

File: jupyter_console/ptshell.py [hunk 1]
```diff
@@ -723,7 +723,10 @@ def handle_image(self, data, mime):
     def handle_image_PIL(self, data, mime):
         if mime not in ('image/png', 'image/jpeg'):
             return False
-        from PIL import Image, ImageShow
+        try:
+            from PIL import Image, ImageShow
+        except ImportError:
+            return False
```
Comment (Carreau) [inline jupyter_console/ptshell.py]: should we `warnings.warn` or `log.warn` something here ?

Comment (flying-sheep) [inline jupyter_console/ptshell.py]: probably [here](https://github.com/flying-sheep/jupyter_console/blob/983b048798f7925d005268cee6a4d9ecd4ac791f/jupyter_console/ptshell.py#L684)

**Comment by Carreau** [comment]
Ok, merging that.

**Pull request merged**